### PR TITLE
Add write ahead log to collector metrics exporter

### PR DESCRIPTION
**Pull request by damemi**
This adds optional WAL buffering to the collector metrics exporter, heavily based on a similar implementation in the [Prometheus remote write exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6f1b6b97eb5d317bee3d20bcf1a8a96ceabdbd6b/exporter/prometheusremotewriteexporter/wal.go).

The flow for this is:
* If WAL is enabled, set it up in the user-specified directory under `gcp_metrics_wal`
* Update `PushMetrics()` to write byte-marshalled CreateTimeSeriesRequest objects to the WAL
    * if WAL is disabled, just export the requests as normal
* Start a `walRunner()` go routine that spawns a `readWALAndExport()` loop (the goroutine also handles shutdown signals)
* `readWALAndExport()` continuously reads the latest entry from the WAL and exports it to GCM:
    * If the export is successful, move the read index forward and continue (truncating the log file)
    * If the export fails for DeadlineExceeded or Unavailable, retry forever until success or non-retryable failure
    * If the export fails for any other reason, move the read index forward and continue (truncating the log file and skipping this entry)
* If there are no entries found in the WAL, `watchWAL()` opens a file watch to block for changes to the log file.

Added an integration test with WAL enabled to test this

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,4 @@ Thumbs.db
 coverage.*
 
 /example/metric/metric
+/exporter/collector/integrationtest/gcp_metrics_wal/
```
Comment (psx95) [inline .gitignore:13]: Nit: Missing newline.

File: exporter/collector/config.go [hunk 2]
```diff
@@ -140,6 +142,18 @@ type MetricConfig struct {
 	EnableSumOfSquaredDeviation bool `mapstructure:"sum_of_squared_deviation"`
 }
 
+// WALConfig defines settings for the write ahead log. WAL buffering writes data
+// points in-order to disk before reading and exporting them. This allows for
```
Comment (jsuereth) [inline exporter/collector/config.go:146]: Is there a way to do in-memory only WAL?

File: exporter/collector/metrics.go
Comment (psx95) [inline exporter/collector/metrics.go:1]: A few newly added lines here are not covered by unit test - some of them involve the specific case when WAL is enabled. Would it be worth adding tests that exercise that logic ?

File: exporter/collector/metrics.go [hunk 3]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+			}
+			// retry at same read index if retryable (network) error
+			s := status.Convert(err)
+			if s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable {
+				me.obs.log.Error("non-retryable error, skipping request")
+				continue
 			}
+
+			err = me.wal.TruncateFront(me.rWALIndex.Load())
 			if err != nil {
```
Comment (jsuereth) [inline exporter/collector/metrics.go:428]: Are you testing `err` to see if it's non-retryable here?

File: exporter/collector/metrics.go [hunk 4]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+			switch event.Op {
+			case fsnotify.Remove:
+				me.obs.log.Error("WAL file deleted")
+			case fsnotify.Rename:
+				me.obs.log.Error("WAL file renamed")
+			case fsnotify.Write:
+				wErr = nil
+			}
+
+		case eerr, ok := <-walWatcher.Errors:
```
Comment (psx95) [inline exporter/collector/metrics.go:492]: Possible typo - `eerr` ?

File: exporter/collector/config.go (hunk 2 above)
Comment (damemi) [reply exporter/collector/config.go:146]: Not with this library: https://github.com/tidwall/wal/blob/e70ddec4a12f90db734d4d90df98b5088bd6d417/wal.go#L159

File: exporter/collector/metrics.go (hunk 3 above)
Comment (damemi) [reply exporter/collector/metrics.go:428]: Retryable errors from GCM are handled in the `continue` statement above (L423). If this line returns an error then it was a problem with the WAL itself

File: exporter/collector/metrics.go
Comment (damemi) [reply exporter/collector/metrics.go:1]: I'll add some unit tests, but just as a note the codecov checks don't know about our integration (fixture) tests, which do cover these lines

File: exporter/collector/metrics.go [hunk 5]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+			if ok {
+				wErr = eerr
+			}
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) walRunner(ctx context.Context) {
```
Comment (psx95) [inline exporter/collector/metrics.go:502]: A comment describing how this function fits in the bigger picture would be helpful to anyone contributing to this file.

File: exporter/collector/metrics.go [hunk 6]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+		select {
+		case <-me.shutdownC:
+			return
+		case <-ctx.Done():
+			wErr = ctx.Err()
+			return
+
+		case event, ok := <-walWatcher.Events:
+			if !ok {
+				return
```
Comment (psx95) [inline exporter/collector/metrics.go:481]: What does `!ok` indicate here ?

File: exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json
Comment (psx95) [inline exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json:1]: The [basic_counter_metrics_expect.json](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/0179cdda02d5c1bb256fa60087883328573f4672/exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_expect.json) has contents differing to this output for the same input. 

No of lines in basic_counter_metrics_expect.json = 693
No of lines in basic_counter_metrics_wal_expect.json = 360

Is this expected ?

Comment (damemi) [reply exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json:1]: That's a good catch, these should be the same. It looks like the test metrics aren't actually being sent to the fixture in this change (only selfObservabilityMetrics). Something's definitely not right with that

File: exporter/collector/metrics.go [hunk 7]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+		// Must have been ErrNotFound, start a file watch and block waiting for updates.
+		if err := me.watchWAL(ctx); err != nil {
+			return err
 		}
 	}
-	if len(errs) > 0 {
-		return multierr.Combine(errs...)
+}
+
+// watchWAL watches the WAL directory for a write then returns to the
```
Comment (punya) [inline exporter/collector/metrics.go:451]: Is this approach commonly used? For example, I don't see it used (at least directly) in the WAL implementation for the PRW exporter.

This would make sense if WAL writes were coming from a different process, but is there a simpler way where we write to a channel after every successful write to the WAL?

File: exporter/collector/metrics.go [hunk 8]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+	return err
+}
+
+func (me *MetricsExporter) walRunner(ctx context.Context) {
+	defer me.goroutines.Done()
+	for {
+		select {
+		case <-me.shutdownC:
+			// do one last final sync/read/export then return
+			err := me.wal.Sync()
```
Comment (punya) [inline exporter/collector/metrics.go:508]: Why is this necessary, given that we've set up our WAL in auto-sync mode?

File: exporter/collector/metrics.go [hunk 9]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) walRunner(ctx context.Context) {
+	defer me.goroutines.Done()
+	for {
+		select {
```
Comment (punya) [inline exporter/collector/metrics.go:505]: Do we need to handle `ctx.Done()`?

File: exporter/collector/metrics.go (hunk 5 above)
Comment (punya) [reply exporter/collector/metrics.go:502]: Also consider giving the function a more descriptive name such as `readAndExportLoop`.

File: exporter/collector/metrics.go [hunk 10]
```diff
@@ -64,14 +71,22 @@ type selfObservability struct {
...
 	// shutdownC is a channel for signaling a graceful shutdown
 	shutdownC chan struct{}
 	// mdCache tracks the metric descriptors that have already been sent to GCM
 	mdCache map[string]*monitoringpb.CreateMetricDescriptorRequest
+	// A channel that receives metric descriptor and sends them to GCM once
+	metricDescriptorC chan *monitoringpb.CreateMetricDescriptorRequest
+	// the "write" index for WAL
+	wWALIndex *atomic.Uint64
+	// the "read" index for WAL
+	rWALIndex *atomic.Uint64
```
Comment (punya) [inline exporter/collector/metrics.go:87]: Does this actually need to be atomic, or could it be a local variable in the read and export loop? It seems like it's only updated in that one goroutine.

**Review by punya — COMMENTED** [review]
Is there a better way to separate the WAL concern from the other parts of our export logic? For example, could you make a self-contained struct that deals with the WAL and has a small API that can be used by the exporter? This could take a couple of forms:

* a struct that has no-op implementations for the nil value (which would remove the repeated nil-checks in the exporter)
* a struct that wraps the "raw" exporter in a version that adds a WAL

File: exporter/collector/metrics.go [hunk 11]
```diff
@@ -272,44 +329,198 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
-			if failedPoints > 0 {
-				recordPointCountDataPoint(ctx, failedPoints, st)
+			err = me.export(ctx, req)
+			if err != nil {
+				me.obs.log.Warn(fmt.Sprintf("error exporting to GCM: %+v", err))
+			}
+			// retry at same read index if retryable (network) error
+			s := status.Convert(err)
+			if s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable {
+				me.obs.log.Error("non-retryable error, skipping request")
```
[the remark this answers is not in the document]
Comment (punya) [reply exporter/collector/metrics.go:423]: Why do we consider DeadlineExceeded and Unavailable to be unrecoverable errors? E.g., it seems like unavailable could be a transient network blip.

Conversely, if we encounter an ostensibly recoverable error but it happens over and over again - should we implement some form of backoff and eventually give up (drop that request)?

File: exporter/collector/metrics.go [hunk 12]
```diff
@@ -191,13 +214,47 @@ func NewGoogleCloudMetricsExporter(
...
+func (me *MetricsExporter) setupWAL() error {
+	walPath := filepath.Join(me.cfg.MetricConfig.WALConfig.Directory, "gcp_metrics_wal")
+	me.walPath = walPath
+	metricWal, err := wal.Open(walPath, nil)
+	if err != nil {
+		return err
+	}
+	me.wal = metricWal
+
+	// sync existing WAL indices
```
Comment (punya) [inline exporter/collector/metrics.go:243]: If there are a nonzero number of entries in the WAL at startup, does it mean we're recovering from a crash/unclean shutdown? In that case it seems useful to log how many entries we're about to try to recover.

(Crash recovery seems like the biggest benefit of file- vs memory-backed WAL.)

File: exporter/collector/metrics.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (punya) [reply exporter/collector/metrics.go:423]: Maybe I'm misunderstanding the control flow, but this logic seems like it contradicts the PR description (i.e. retryable vs non-retryable seem flipped).

File: exporter/collector/config.go [hunk 13]
```diff
@@ -103,7 +98,14 @@ type MetricConfig struct {
 	// ExtraMetrics is an extension point for exporters to add to the set
 	// of ResourceMetrics during a call to PushMetrics.
 	ExtraMetrics func(pmetric.Metrics) pmetric.ResourceMetricsSlice
-	Prefix       string `mapstructure:"prefix"`
+	// GetMetricName is not settable in config files, but can be used by other
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/config.go:101]: these are caught by the `fieldalignment` filter, and won't pass CI without the formatting changes in this PR

File: exporter/collector/config.go (hunk 2 above)
Comment (jsuereth) [reply exporter/collector/config.go:146]: I think handling network outages is the priority.  5-min outage with even a potential collector crash/reboot would be handled by the on-disk portion here.

In-memory is a nice to have, unless the WAL is not viable for some reason.

File: exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json
Comment (damemi) [reply exporter/collector/integrationtest/testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json:1]: added cross-reference option to testcases

**Comment by damemi** [comment]
Sorry for the push spam, trying to debug some flakey CI test results that I can't reproduce as reliably locally or in a container

File: internal/cloudmock/metrics.go [hunk 14]
```diff
@@ -104,10 +105,16 @@ func (f *fakeMetricServiceServer) CreateTimeSeries(
 	var code codes.Code
 	if strings.Contains(req.Name, "notfound") {
 		code = codes.NotFound
+	} else if strings.Contains(req.Name, "unavailable") {
```
Comment (jsuereth) [inline internal/cloudmock/metrics.go:108]: Please add docs somewhere denoting these magic strings.

File: exporter/collector/config.go [hunk 15]
```diff
@@ -140,6 +142,18 @@ type MetricConfig struct {
...
+// WALConfig defines settings for the write ahead log. WAL buffering writes data
+// points in-order to disk before reading and exporting them. This allows for
+// better retry logic when exporting fails (such as a network outage), because
+// it preserves both the data on disk and the order of the data points.
+type WALConfig struct {
+	// Directory is the location to store WAL files.
+	Directory string `mapstructure:"directory"`
+	// Enabled turns the WAL on or off. When false, the exporter will not use
+	// the WAL and will instead attempt to send data directly to Google Cloud.
+	Enabled bool `mapstructure:"enabled"`
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/config.go:156]: changed this to a pointer and check for nil when instantiating

File: exporter/collector/metrics.go [hunk 16]
```diff
@@ -116,6 +131,12 @@ func (me *MetricsExporter) Shutdown(ctx context.Context) error {
 	go func() {
 		// Wait until all goroutines are done
 		me.goroutines.Wait()
+		// Close the WAL if open
+		if me.wal != nil {
+			if err := me.wal.Close(); err != nil {
+				fmt.Printf("error closing WAL: %+v\n", err)
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/metrics.go:138]: done

File: exporter/collector/metrics.go (hunk 4 above)
Comment (damemi) [reply exporter/collector/metrics.go:582]: this is intentionally `eerr` to avoid duplicating the `err` variable above ... renamed to be better

File: exporter/collector/metrics.go (hunk 10 above)
Comment (damemi) [reply exporter/collector/metrics.go]: removed these

File: exporter/collector/metrics.go (hunk 8 above)
Comment (damemi) [reply exporter/collector/metrics.go]: removed

File: exporter/collector/metrics.go [hunk 17]
```diff
@@ -272,44 +360,252 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+			}
+			// ErrNotFound from wal.Read() means the index is either 0 or out of
+			// bounds (indicating we're probably at the end of the WAL). That error
+			// will trigger a file watch for new writes (below this). For other
+			// errors, fail.
+			if !errors.Is(err, wal.ErrNotFound) {
+				me.obs.log.Error(fmt.Sprintf("error reading WAL and exporting: %+v", err))
+			}
+
+			// Must have been ErrNotFound, start a file watch and block waiting for updates.
```
Comment (ridwanmsharif) [inline exporter/collector/metrics.go:603]: The above `if` block doesn't break or return. Won't we get here even if `err` is another error?

File: exporter/collector/metrics.go [hunk 18]
```diff
@@ -272,44 +360,252 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
-			st = statusCodeToString(s)
-
-			succeededPoints := len(ts)
-			failedPoints := 0
-			for _, detail := range s.Details() {
-				if summary, ok := detail.(*monitoringpb.CreateTimeSeriesSummary); ok {
-					failedPoints = int(summary.TotalPointCount - summary.SuccessPointCount)
-					succeededPoints = int(summary.SuccessPointCount)
-				}
+			if !(s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable) {
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/metrics.go:475]: Yes, but also to keep the ordering of the requests

File: exporter/collector/config.go [hunk 19]
```diff
@@ -140,6 +142,17 @@ type MetricConfig struct {
 	EnableSumOfSquaredDeviation bool `mapstructure:"sum_of_squared_deviation"`
 }
 
+// WALConfig defines settings for the write ahead log. WAL buffering writes data
+// points in-order to disk before reading and exporting them. This allows for
+// better retry logic when exporting fails (such as a network outage), because
+// it preserves both the data on disk and the order of the data points.
+type WALConfig struct {
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/config.go:149]: Agreed, but I think that is fine as a follow up feature request

File: exporter/collector/metrics.go [hunk 20]
```diff
@@ -191,15 +215,79 @@ func NewGoogleCloudMetricsExporter(
...
+	walPath := filepath.Join(me.cfg.MetricConfig.WALConfig.Directory, "gcp_metrics_wal")
+	me.wal.path = walPath
+	metricWal, err := wal.Open(walPath, &wal.Options{LogFormat: 1})
+	if err != nil {
+		return 0, 0, err
+	}
+	me.wal.Log = metricWal
+
+	// default to 1 hour exponential backoff
+	me.wal.maxBackoff = time.Duration(3600 * time.Second)
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/metrics.go]: done

File: exporter/collector/metrics.go (hunk 17 above)
Comment (damemi) [reply exporter/collector/metrics.go:609]: The idea was just to log errors that aren't `ErrNotFound`, since not found means we're at the end of the WAL and can be expected sometimes. added a comment explaining this

File: internal/cloudmock/metrics.go (hunk 14 above)
Comment (damemi) [reply internal/cloudmock/metrics.go]: These are only used in this internal mock server for testing. But I added a comment for this function explaining that a lot better

File: exporter/collector/README.md [hunk 21]
```diff
@@ -178,7 +178,11 @@ Note: These `retry_on_failure` and `sending_queue` are provided (and documented)
 Additional configuration for the metric exporter:
 
 - `metric.prefix` (optional): MetricPrefix overrides the prefix / namespace of the Google Cloud metric type identifier. If not set, defaults to "custom.googleapis.com/opencensus/"
-- `metric.skip_create_descriptor` (optional): Whether to skip creating the metric descriptor.
+- `metric.skip_create_descriptor` (optional): Whether to skip creating the
+  metric descriptor.
+- `metric.wal.directory` (optional): Path to local write-ahead-log file.
+- `metric.wal.max_backoff` (optional): Maximum duration to retry entries from
```
[the remark this answers is not in the document]
Comment (punya) [reply exporter/collector/README.md:184]: I recall that in the collector, there's a convention of using an `experimental_` prefix to signal such things in configuration.

Do we think that this is going to be replaced with an enhanced version of `queued_retry` in future?

File: exporter/collector/config.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (punya) [reply exporter/collector/config.go:149]: Filed #646

File: exporter/collector/integrationtest/testcases/testcases_metrics.go [hunk 22]
```diff
@@ -229,5 +231,78 @@ var MetricsTestCases = []TestCase{
 			cfg.MetricConfig.ClientConfig.Compression = "gzip"
 		},
 	},
+	{
+		Name:                 "Write ahead log enabled",
+		OTLPInputFixturePath: "testdata/fixtures/metrics/basic_counter_metrics.json",
+		ExpectFixturePath:    "testdata/fixtures/metrics/basic_counter_metrics_wal_expect.json",
+		CompareFixturePath:   "testdata/fixtures/metrics/basic_counter_metrics_expect.json",
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/integrationtest/testcases/testcases_metrics.go:238]: "expect" is what the input for a test converts into, and it verifies that a single test hasn't changed since its last fixture was checked in

"compare" is to cross-reference output fixtures between different tests that should have equal outcomes. In this case, we want to make sure that enabling WAL gives the same output as normal export, based on the same input fixture.

File: exporter/collector/integrationtest/inmemoryocexporter.go [hunk 23]
```diff
@@ -162,6 +162,7 @@ func NewMetricTestExporter(
 		ctx,
 		cfg,
 		zap.NewNop(),
+		//zap.NewExample(),
```
Comment (punya) [inline exporter/collector/integrationtest/inmemoryocexporter.go:165]:
```suggestion
```

File: exporter/collector/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/README.md:184]: I'm fine with that, but our exporter is beta so I don't think we can change these settings once we publish them

File: exporter/collector/metrics.go (hunk 9 above)
Comment (aabmass) [reply exporter/collector/metrics.go:579]: It looks like this context is the one passed to `NewGoogleCloudMetricsExporter()`. Is that one intended for initialization only? We're not even passing it to `go mExp.exportMetricDescriptorRunner()` on L235.

File: exporter/collector/metrics.go [hunk 24]
```diff
@@ -271,44 +365,254 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
...
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) runWALReadAndExportLoop(ctx context.Context) {
+	defer me.goroutines.Done()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
Comment (aabmass) [inline exporter/collector/metrics.go:577]: We're not using this cancel anywhere, I think this code is a noop

File: exporter/collector/metrics.go [hunk 25]
```diff
@@ -115,6 +138,12 @@ func (me *MetricsExporter) Shutdown(ctx context.Context) error {
 	go func() {
 		// Wait until all goroutines are done
 		me.goroutines.Wait()
+		// Close the WAL if open
+		if me.wal != nil {
+			if err := me.wal.Close(); err != nil {
+				me.obs.log.Error(fmt.Sprintf("error closing WAL: %+v\n", err))
+			}
```
Comment (aabmass) [inline exporter/collector/metrics.go:145]: Consider moving this into into a defer in `runWALReadAndExportLoop()` so the goroutine can handle its own cleanup when it receives the shutdown signal.

**Review by aabmass — APPROVED** [review]
I think it might be good to follow up PR with refactoring the WAL into its own struct as `metrics.go` is getting pretty long and I think the sprinkled WAL logic could be encapsulated

File: exporter/collector/metrics.go [hunk 26]
```diff
@@ -182,6 +211,7 @@ func NewGoogleCloudMetricsExporter(
 		shutdownC:         shutdown,
 		timeout:           timeout,
 	}
+	mExp.exportFunc = mExp.exportToTimeSeries
```
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/metrics.go:214]: `exportToTimeSeries()` is a method on a MetricsExporter struct, so I don't think I can set it in the initializer

File: exporter/collector/metrics.go (hunk 9 above)
Comment (damemi) [reply exporter/collector/metrics.go:579]: `shutdownC` is only closed if `Shutdown()` is called, so I think handling the main context closing is good to have to prevent from blocking on other terminations (non-gracefully?)

File: exporter/collector/metrics.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (aabmass) [reply exporter/collector/metrics.go:214]: Ah that makes sense

File: exporter/collector/metrics.go (hunk 25 above)
Comment (damemi) [reply exporter/collector/metrics.go:145]: sgtm

File: exporter/collector/metrics.go (hunk 24 above)
Comment (damemi) [reply exporter/collector/metrics.go]: sounds good, removed this and changed the `runCtx` to just `ctx`

File: exporter/collector/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/README.md]: ah yeah I commented without seeing Punya's update

I changed these to add the `experimental_` prefix, and also to the wal object... it looks a little weird like this:

```yaml
metric:
  experimental_wal:
    experimental_directory: my/path
```

do you think we could just make the `wal` subsection experimental and imply that everything under it is also subject to change? Or just to be safe, mark each option experimental?

```yaml
metric:
  experimental_wal:
    directory: my/path #subject to change
```

[the remark this answers is not in the document]
Comment (aabmass) [reply exporter/collector/README.md]: I think the second less verbose option is fine. I see they are doing that [this processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7c0fd3de2aa7cbf61d6360c6208079c761992512/processor/metricsgenerationprocessor/README.md#configuration) as well.

File: exporter/collector/metrics.go (hunk 9 above)
Comment (aabmass) [reply exporter/collector/metrics.go:577]: > so I think handling the main context closing is good to have to prevent from blocking on other terminations (non-gracefully?)

That makes sense, but I wasn't sure of the collector's contract here. I figured it would eventually cancel that "initialization" context which would cause this goroutine to exit but that doesn't seem to be the case.

File: exporter/collector/integrationtest/testcases/testcases_metrics.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (damemi) [reply exporter/collector/integrationtest/testcases/testcases_metrics.go:249]: there is also a comment in the definition in testcase.go:

```
 	// CompareFixturePath is a second output fixture that should be equal to this test's output fixture.
	// Used for cross-referencing multiple tests that should have the same output, without overwriting the same fixtures.
```

**Comment by damemi** [comment]
I think I've addressed all the blockers, going to merge this now so we can get an initial release with it out.
